CMakeLists.txt

```diff
@@ -1,170 +1,89 @@
-
-option(RESULT_COMPILE_UNIT_TESTS "Compile and run the unit tests for this library" OFF)
-
-if (NOT CMAKE_TESTING_ENABLED AND RESULT_COMPILE_UNIT_TESTS)
-  enable_testing()
-endif ()
-
-project(Result
-  VERSION "1.0.0"
-  LANGUAGES CXX
+cmake_minimum_required(VERSION 3.14)
+
+project(
+    Result
+    VERSION 1.0.0
+    DESCRIPTION "A lightweight C++11-compatible error-handling mechanism"
+    HOMEPAGE_URL "https://github.com/bitwizeshift/result"
+    LANGUAGES CXX
 )
 
-set(RESULT_VERSION_MAJOR ${PROJECT_VERSION_MAJOR} CACHE INTERNAL "Major version of Result")
-set(RESULT_VERSION_MINOR ${PROJECT_VERSION_MINOR} CACHE INTERNAL "Minor version of Result")
-set(RESULT_VERSION_PATCH ${PROJECT_VERSION_PATCH} CACHE INTERNAL "Patch version of Result")
-set(RESULT_VERSION ${PROJECT_VERSION} CACHE INTERNAL "Version of Result")
-
-if (EXISTS "${CMAKE_CURRENT_BINARY_DIR}/conanbuildinfo.cmake")
-  include("${CMAKE_CURRENT_BINARY_DIR}/conanbuildinfo.cmake")
-  conan_set_find_paths()
-endif ()
-
-##############################################################################
-# Targets
-##############################################################################
-
-set(CMAKE_CXX_STANDARD 11)
-set(CMAKE_CXX_STANDARD_REQUIRED ON)
-set(CMAKE_CXX_EXTENSIONS OFF)
-
-set(header_files
-  include/result.hpp
-)
-
-add_library(${PROJECT_NAME} INTERFACE)
-add_library(${PROJECT_NAME}::${PROJECT_NAME} ALIAS ${PROJECT_NAME})
-
-target_include_directories(${PROJECT_NAME}
-  INTERFACE $<BUILD_INTERFACE:${CMAKE_CURRENT_LIST_DIR}/include>
-  INTERFACE $<INSTALL_INTERFACE:include>
+string(
+    COMPARE EQUAL
+    "${CMAKE_SOURCE_DIR}" "${CMAKE_CURRENT_SOURCE_DIR}"
+    is_top_project
 )
 
-if ("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang" AND
-    "${CMAKE_CXX_SIMULATE_ID}" STREQUAL "MSVC")
-  # clang-cl does not appear to implement '-pedantic' or 'pedantic-errors',
-  # so this case needs to be handled specifically
-  add_compile_options(-Wall -Werror -Wextra)
-
-  # clang-cl treats '-Wall' like '-Weverything' currently; so there are a few
-  # warnings we need to manually disable.
-
-  # Disable unnecessary compatibility and 'newline-eof'. This is a modern C++
-  # library -- so these serve no purpose
-  add_compile_options(-Wno-c++98-compat -Wno-c++98-compat-pedantic -Wno-newline-eof)
-
-  # Disable warnings on padding. This is a great warning, but not good when
-  # coupled with -Werror
-  add_compile_options(-Wno-padded)
-
-  # Several utilities require static objects that have exit-time destructors,
-  # including 'error_category' types from the standard.
-  add_compile_options(-Wno-exit-time-destructors)
+# ---- Warning guard ----
 
-  # This is erroring on valid uses of the standard library, like std::lock_guard.
-  add_compile_options(-Wno-ctad-maybe-unsupported)
+# Protect dependents from this project's warnings if the guard isn't disabled
+set(Result_warning_guard SYSTEM)
+if(is_top_project OR Result_INCLUDE_WITHOUT_SYSTEM)
+  set(Result_warning_guard "")
+endif()
```

Comment:
`IMPORTED` targets already provide implicitly `SYSTEM` include paths. This opt-in `SYSTEM` merely requests the compiler switch that silences warnings, nothing more.

The purpose of the warning guard is to mirror the behavior of `IMPORTED` targets that a user gets from `find_package`.
```cmake
# Doesn't matter how you import a project
add_subdirectory(dependency)
# or
find_package(dependency)

# You will not have to deal with the dependency's warnings either way
target_link_library(my_project_target PRIVATE dependency::target)
```
This above is almost exactly the same reason why you provide an `ALIAS` target as well, to provide the same build and install interfaces for dependents.

Reply:
> `IMPORTED` targets already provide implicitly `SYSTEM` include paths

Do you have any documentation for this claim? I've never heard this as any guaranteed, and even [`add_library`](https://cmake.org/cmake/help/latest/command/add_library.html#id5) doesn't indicate that `IMPORTED` implies `SYSTEM`. As far as I'm aware, this isn't any sort of guarantee that CMake provides, as it entirely depends on how the targets were exported.

My understanding is that the `*Targets.cmake` file defines the `IMPORTED` target exactly as exported -- which would include either [`INTERFACE_INCLUDE_PATHS`](https://cmake.org/cmake/help/latest/prop_tgt/INTERFACE_INCLUDE_DIRECTORIES.html) or [`INTERFACE_SYSTEM_INCLUDE_PATHS`](https://cmake.org/cmake/help/latest/prop_tgt/INTERFACE_SYSTEM_INCLUDE_DIRECTORIES.html) depending on how it was set. I've never seen any documentation or claims that one somehow transforms into the other due to an export.

Reply:
Ah, after searching I finally came across the [`NO_SYSTEM_FROM_IMPORTED`](https://cmake.org/cmake/help/latest/prop_tgt/NO_SYSTEM_FROM_IMPORTED.html) which is the only real reference that seems to indicate this.

TBH though, since the `SYSTEM` property doesn't really do much with compilers like MSVC (it just produces the same `/I`), I'm more inclined to consider setting this property on the targets before exporting so that this implicit `SYSTEM` conversion doesn't happen.
I'd much rather have errors reported consistently across *all* consumers, regardless of build system -- whether it's makefile, CMake, etc. This way users who copy/paste, download a release, etc, won't mysteriously break whereas others work without issue.

Reply:
MSVC is usually the odd one out among the compilers, so this can be chalked up to that again.

You mention consistency, so I say that on platforms where `SYSTEM` works as it should, it's good to provide the exact same experience like the code block above indicates.